File: Tools/Logging.cs
using System;
using System.IO;
using System.Linq;
using System.Text;
using System.Collections.Generic;

using Android.App;
using static Android.Media.Audiofx.DynamicsProcessing;

namespace VorratsUebersicht
{
    internal class Logging
    {
        internal static string GetCurrentLogFileName(DateTime? day = null)
        {
            if (day == null)
            {
                day = DateTime.Today;
            }

            // "/data/user/0/de.stryi.Vorratsuebersicht/files"
            string logFilePath = System.Environment.GetFolderPath(System.Environment.SpecialFolder.Personal);

            // 1. "/storage/emulated/0/Android/data/de.stryi.Vorratsuebersicht/cache"
            // 2. "/storage/1820-3B0F/Android/data/de.stryi.Vorratsuebersicht/cache"
            var cacheDirs = Application.Context.GetExternalCacheDirs();
            if (cacheDirs.Length > 1)
            {
                logFilePath = cacheDirs[0].AbsolutePath;
            }

            string fileName = string.Format("Vue_{0}.log", day.Value.ToString("yyyy.MM.dd"));
            return Path.Combine(logFilePath, fileName);
        }

        internal static void CreateTestLogFiles()
        {
            for (int index = -1; index >= -11; index--)
            {
                var day = DateTime.Today.AddDays(index);
                string fileName = Logging.GetCurrentLogFileName(day);

                if (!File.Exists(fileName))
                {
                    File.AppendAllText(fileName, DateTime.Now.ToString("yyyy-MM-dd HH.mm.ss - ") + "Protokolleintrag 1\n");
                    File.AppendAllText(fileName, DateTime.Now.ToString("yyyy-MM-dd HH.mm.ss - ") + "Protokolleintrag 2\n");
                    File.AppendAllText(fileName, DateTime.Now.ToString("yyyy-MM-dd HH.mm.ss - ") + "Protokolleintrag 3\n");
                }
            }
        }

        /// <summary>
        /// Nur die letzten 5 Log Dateien behalten (von den letzten 5 Tagen)
        /// </summary>
        internal static void ClearOldLogFiles()
        {
            //Logging.CreateTestLogFiles();

            var fileList = Logging.GetLogFileList();
            if (fileList.Length <= 5)
                return;

            for (int i = 0; i < fileList.Length - 5; i++)
            {
                File.Delete(fileList[i]);
            }
        }

        internal static string[] GetLogFileList()
        {
            // "/data/user/0/de.stryi.Vorratsuebersicht/files"
            string logFilePath = System.Environment.GetFolderPath (System.Environment.SpecialFolder.Personal);
            
            var unsortedFileList = Directory.GetFiles(logFilePath, "Vue_*.log");

            List<string> sortedList = new List<string>(unsortedFileList);

            // 1. "/storage/emulated/0/Android/data/de.stryi.Vorratsuebersicht/cache"
            // 2. "/storage/1820-3B0F/Android/data/de.stryi.Vorratsuebersicht/cache"
            var cacheDirs = Application.Context.GetExternalCacheDirs();
            if (cacheDirs.Length > 1)
            {
                logFilePath = cacheDirs[0].AbsolutePath;

                unsortedFileList = Directory.GetFiles(logFilePath, "Vue_*.log");
                sortedList.AddRange(unsortedFileList);
            }

            return sortedList.OrderBy(e => e).ToArray();
        }

        internal static string GetLogFileText()
        {
            StringBuilder text = new StringBuilder();

            var fileList = Logging.GetLogFileList();

            foreach(string logFileName in fileList)
            {
                text.AppendLine("LogFile: " + logFileName);
                text.AppendLine(File.ReadAllText(logFileName));
            }

            return text.ToString();
        }

    }
}

File: app/src/main/java/de/stryi/vorratsuebersicht/tools/Logging.kt
package de.stryi.vorratsuebersicht.tools

import android.content.Context
import android.util.Log
import java.io.File
import java.lang.StringBuilder
import java.text.SimpleDateFormat
import java.util.*

object Logging {

    var logFileName : String = ""

    private val dateFormatDay = SimpleDateFormat("yyyy.MM.dd", Locale.getDefault())
    private val dateFormatLog = SimpleDateFormat("yyyy-MM-dd HH.mm.ss - ", Locale.getDefault())

    fun InitializeLogFile(context: Context)
    {
        // Beispiel: /storage/emulated/0/Android/data/de.stryi.vorratsuebersicht2/cache/Vue_2026.08.10.log
        logFileName = getCurrentLogFileName(context)

        deleteOldLogFiles(context)
    }

    fun getCurrentLogFileName(context: Context, day: Date? = null): String {

        val usedDay = day ?: Date()

        // "/data/user/0/<package>/files"
        var logFilePath = context.filesDir.absolutePath

        // externe Cache-Verzeichnisse
        val cacheDirs = context.externalCacheDirs
        if (!cacheDirs.isNullOrEmpty()) {
            logFilePath = cacheDirs[0].absolutePath
        }

        val fileName = "Vue_${dateFormatDay.format(usedDay)}.log"
        return File(logFilePath, fileName).absolutePath
    }

    fun logToFile(text: String) {

        // Beispiel: /storage/emulated/0/Android/data/de.stryi.vorratsuebersicht2/cache/Vue_2026.08.10.log
        if (logFileName.isEmpty())
            return

        try {
            File(logFileName).appendText(text)
        }
        catch (e: Exception) {
            Log.e("stryi", e.toString())
        }
    }

    fun createTestLogFiles(context: Context) {

        val calendar = Calendar.getInstance()

        for (index in -1 downTo -11) {

            val cal = calendar.clone() as Calendar
            cal.add(Calendar.DAY_OF_MONTH, index)

            val fileName = getCurrentLogFileName(context, cal.time)
            val file = File(fileName)

            if (!file.exists()) {

                file.appendText(dateFormatLog.format(Date()) + "Protokolleintrag 1\n")
                file.appendText(dateFormatLog.format(Date()) + "Protokolleintrag 2\n")
                file.appendText(dateFormatLog.format(Date()) + "Protokolleintrag 3\n")
            }
        }
    }

    /**
     * Nur die letzten 5 Logdateien behalten
     */
    fun deleteOldLogFiles(context: Context) {

        val fileList = getLogFileList(context)
        if (fileList.size <= 5) return

        for (i in 0 until fileList.size - 5) {
            File(fileList[i]).delete()
        }
    }

    fun getLogFileList(context: Context): Array<String> {

        // Beispiel: /data/user/0/de.stryi.vorratsuebersicht2/files
        var logFilePath = context.filesDir.absolutePath

        val sortedList = mutableListOf<String>()

        val baseDir = File(logFilePath)
        val unsorted = baseDir.listFiles { _, name ->
            name.startsWith("Vue_") && name.endsWith(".log")
        }

        unsorted?.forEach { sortedList.add(it.absolutePath) }

        // Beispiel: "/storage/emulated/0/Android/data/de.stryi.vorratsuebersicht2/cache"
        //           "/storage/0000-0000/Android/data/de.stryi.vorratsuebersicht2/cache"
        // Wenn keine SD Karte, dann nur der erste Eintrag.
        val cacheDirs = context.externalCacheDirs
        if (cacheDirs != null && cacheDirs.size > 0) {

            logFilePath = cacheDirs[0].absolutePath
            val cacheDir = File(logFilePath)

            val moreFiles = cacheDir.listFiles { _, name ->
                name.startsWith("Vue_") && name.endsWith(".log")
            }

            moreFiles?.forEach { sortedList.add(it.absolutePath) }
        }

        return sortedList.sorted().toTypedArray()
    }

    fun getLogFileText(context: Context): String {

        val text = StringBuilder()
        val fileList = getLogFileList(context)

        for (logFileName in fileList) {

            text.appendLine("LogFile: $logFileName")

            val file = File(logFileName)
            if (file.exists()) {
                text.appendLine(file.readText())
            }

            text.appendLine("**** E N D   OF    F I L E ****")
            text.appendLine()
        }

        return text.toString()
    }

    fun deleteAllLogFiles(context: Context) {
        val fileList = getLogFileList(context)

        for (logFileName in fileList)
        {
            File(logFileName).delete()
        }
    }
}